docs/designers-developers/developers/backward-compatibility/README.md

```diff
@@ -1 +1,65 @@
 # Backward Compatibility
+
+Historically, WordPress has been known for preserving backward compatibility across versions. Gutenberg follows this example wherever possible in its production public APIs. There are rare occasions where breaking backward compatibility is unavoidable and in those cases the breakage:
+
+* Should be constrained as much as possible to a small surface area of the API.
+* Should be documented as clearly as possible to third-party developers using Dev Notes.
+
+## What qualifies as a production public API
+
+The Gutenberg code base is composed of two different types of packages: 
+ - **production packages**: these are packages that are shipped as WordPress scripts (example: wp-components, wp-editor...).
+ - **development packages**: these are made up of developer tools that can be used by third-party developers to lint, test, format and build their themes and plugins (example: @wordpress/scrips, @wordpress/env...). Typically, these are consumed as npm dependencies in third-party projects.
+
+Backward compatibility guarantees only apply to the production packages, as updates happen through WordPress upgrades.
+ 
+Production packages use the `wp` global variable to provide APIs to third-party developers. These APIs can be JavaScript functions, variables and React components.
+
+### How to preserve backward compatibility for a JavaScript function
+
+* The name of the function should not change.
+* The order of the arguments of the function should not change.
+* The function's returned value type should not change.
+* Changes to arguments (new arguments, modification of semantics) is possible if we guarantee that all previous calls are still possible.
+
+### How to preserve backward compatibility for a React Component
+
+* The name of the component should not change.
+* The props of the component should not be removed.
+* Existing prop values should continue to be supported. If a component accepts a function as a prop, we can update the component to accept a new type for the same prop, but it shouldn't break existing usage.
+* Adding new props is allowed.
+* React Context dependencies can only be added or removed if we ensure the previous context contract is not breaking.
+
+### How to preserve backward compatibility for a Block
+
+* Existing usage of the block should not break or be marked as invalid when the editor is loaded.
+* The styling of the existing blocks should be guaranteed.
+* Markup changes should be limited to the minimum possible, but if a block needs to change its saved markup, making previous versions invalid, a [**deprecated version**](/docs/designers-developers/developers/block-api/block-deprecation.md) of the block should be added.
+
+## Class names and DOM updates
+
+Class names and DOM nodes used inside the tree of React components are not considered part of the public API and can be modified. 
```

Comment:
One fundamental concern I have here (and apologies for the late revisit) is that currently it is _only_ possible to build a block which renders child blocks horizontally by targeting class names, as the `InnerBlocks` component does not offer declarative UI configuration (e.g. `isHorizontal` or `shouldWrapBlocks`-style props).

I love the idea here but it feels like this is a case where it might be warranted to make an exception until a declarative alternative is available.

Reply:
That's a good point, I think the Navigation block has started building a declarative API for that, the columns block is also a horizontal container, it should be possible to offer a better API here.

Reply:
I don't think this example contradicts the general recommendation though. If there's need for something like this, we shouldn't want that the developer feel need to target the classes, but instead offer it as an option via the component interface. It appears that's what's already being said in this comment chain, but I want to point out that I think it's consistent with treating the class names as unstable.